tsdb/store.go

```diff
@@ -431,27 +457,32 @@ func (s *Store) loadShards() error {
 		return nil
 	}
 
-	// We use `rawShardCount` as a buffer size for channel creation below.
-	// If there is no startupProgressMetrics count then this will be 0 creating a
-	// zero buffer channel.
-	rawShardCount := 0
-	if s.startupProgressMetrics != nil {
-		err := walkShardsAndProcess(func(sfile *SeriesFile, idx interface{}, sh os.DirEntry, db os.DirEntry, rp os.DirEntry) error {
-			rawShardCount++
+	// Find total shard count for progress and channel size.
+	totalShardCount := 0
+	err = walkShardsAndProcess(func(shardID uint64, sh os.DirEntry, db os.DirEntry, rp os.DirEntry) error {
+		totalShardCount++
+		if s.startupProgressMetrics != nil {
 			s.startupProgressMetrics.AddShard()
-			return nil
-		})
-		if err != nil {
-			return err
 		}
+		return nil
+	})
+	if err != nil {
+		return err
 	}
 
-	shardResC := make(chan *shardResponse, rawShardCount)
-	err = walkShardsAndProcess(func(sfile *SeriesFile, idx interface{}, sh os.DirEntry, db os.DirEntry, rp os.DirEntry) error {
-		shardLoaderWg.Add(1)
+	// Do the actual work of loading shards.
+	shardResC := make(chan *shardResponse, totalShardCount)
+	pendingShardCount := 0
+	err = walkShardsAndProcess(func(shardID uint64, sh os.DirEntry, db os.DirEntry, rp os.DirEntry) error {
+		pendingShardCount++
 
-		go func(db, rp, sh string) {
-			defer shardLoaderWg.Done()
+		loadSingleShard := func(shardID uint64, db, rp, sh string) (rshard *Shard, rerr error) {
```

Comment:
For cleaning up code, I would prefer `loadSingleShard` to be a private method of the `Store` that took the `Limiter` `t`  and the channel `shardResC` as arguments from a little wrapper lambda, rather than defining `loadSingleShard` inline as a huge lambda.  But this is an aesthetic argument, nothing more.

Reply:
The limiter and channel are pretty closely coupled with the process of loading all the shards on disk in a single pass. There is a large amount of duplicate code to be found in `Store.CreateShard`, but the limiter and channel absolutely don't apply there. The mutex prevents concurrent calls to `CreateShard`. The code in `loadShards` lives on the edge with how it handles loading the shards in parallel. It's correct and that looks intentional, although I would have preferred additional comments.

I'll try to break out the common code pieces between `loadShards` and `CreateShard`.